pkg/resource/providerconfig.go

```diff
@@ -46,6 +55,50 @@ func IsMissingReference(err error) bool {
 	return ok
 }
 
+// EnvLookupFn looks up an environment variable.
+type EnvLookupFn func(string) string
+
+// ExtractEnv extracts credentials from an environment variable.
+func ExtractEnv(ctx context.Context, e EnvLookupFn, s xpv1.CommonCredentialSelectors) ([]byte, error) {
+	if s.Env == nil {
+		return nil, errors.New(errExtractEnv)
+	}
+	return []byte(e(s.Env.Name)), nil
+}
+
+// ExtractFs extracts credentials from the filesystem.
+func ExtractFs(ctx context.Context, fs afero.Fs, s xpv1.CommonCredentialSelectors) ([]byte, error) {
+	if s.Fs == nil {
+		return nil, errors.New(errExtractFs)
+	}
+	return afero.ReadFile(fs, s.Fs.Path)
+}
+
+// ExtractSecret extracts credentials from a Kubernetes secret.
+func ExtractSecret(ctx context.Context, client client.Client, s xpv1.CommonCredentialSelectors) ([]byte, error) {
+	if s.SecretRef == nil {
+		return nil, errors.New(errExtractSecretKey)
+	}
+	secret := &corev1.Secret{}
+	if err := client.Get(ctx, types.NamespacedName{Namespace: s.SecretRef.Namespace, Name: s.SecretRef.Name}, secret); err != nil {
+		return nil, errors.Wrap(err, errGetCredentialsSecret)
+	}
+	return secret.Data[s.SecretRef.Key], nil
+}
+
+// CommonCredentialExtractor extracts credentials from common sources.
+func CommonCredentialExtractor(ctx context.Context, source xpv1.CredentialsSource, client client.Client, selector xpv1.CommonCredentialSelectors) ([]byte, error) {
+	switch source { // nolint:exhaustive
+	case xpv1.CredentialsSourceEnvironment:
+		return ExtractEnv(ctx, os.Getenv, selector)
+	case xpv1.CredentialsSourceFilesystem:
+		return ExtractFs(ctx, afero.NewOsFs(), selector)
+	case xpv1.CredentialsSourceSecret:
+		return ExtractSecret(ctx, client, selector)
+	}
```

Comment:
Should we add a case for `None` that just does nothing (i.e. doesn't return creds or an error)?

Reply:
I was a little concerned with whether we would want to return `nil` bytes with `nil` error if users were just passing that along to some sort of client constructor. They could always override though so it is likely fine.